File: bindings/python/vizstack/assembler.py
import json
from typing import Any, Optional, List, MutableSet, Dict, Union

from vizstack.types import JsonType
from vizstack.view import get_view, View


__all__ = ['assemble']


def assemble(obj: Any) -> Dict[str, Union[str, Dict[str, Dict[str, JsonType]]]]:
    """Returns a dict with all of the information needed to render a visualization of `obj`.

    Args:
        obj: An object to be visualized.

    Returns:
        {
            'models': Maps a View ID to its ViewModel.
            'rootId': the View ID of the top-level View in 'models'.
        }
    """
    to_add: List[View] = [get_view(obj)]
    added: MutableSet[str] = set()
    return_dict: Dict[str, Union[Optional[str], Dict[str, JsonType]]] = {
        'rootId': to_add[0].id,
        'models': dict(),
    }
    while len(to_add) > 0:
        # Create the ViewModel, add it to `return_dict`, then enqueue all Views referenced in the ViewModel
        view_obj: View = to_add.pop()
        view_id: str = view_obj.id
        if view_id in added:
            continue
        added.add(view_id)
        return_dict['models'][view_id], referenced_views = view_obj.assemble()
        to_add += referenced_views
    return return_dict

File: bindings/javascript/vizstack/src/assembler.js
// @flow
// TODO: document this. Until then, look at Python bindings, since this closely mirrors them.
import { View, _getView } from './view';

// TODO: We can't type this until we can import stuff from `schema.js` in core, but that would create a cyclic dependency
export function assemble(obj: any): {
    rootId: string,
    models: any,
} {
    const toAdd: Array<View> = [_getView(obj)];
    const added: Set<string> = new Set();
    const returnDict: {rootId: string, models: any} = {
        rootId: toAdd[0].id,
        models: {},
    };
    while (toAdd.length > 0) {
        const view = toAdd.pop();
        const viewId = view.id;
        if (added.has(viewId)) {
            continue;
        }
        added.add(viewId);
        const { viewModel, referencedViews } = view.assemble();
        returnDict.models[viewId] = viewModel;
        toAdd.push(...referencedViews);
    }
    return returnDict;
}